python/ray/data/_internal/batcher.py

```diff
@@ -316,5 +374,113 @@ def next_batch(self) -> Block:
+                more computation overlap for very expensive downstream udfs.
+                However it comes at the cost of additional memory overhead.
+            fetch_timeout_s: Seconds to wait while fetching the next batch before
+                timing out.
+        """
+        self.base_batcher = base_batcher
+        self.fetch_timeout_s = fetch_timeout_s
+        self.buffer = queue.Queue(maxsize=buffer_max_size)
+        self.is_thread_running = False
+
+    def add(self, block: Block):
+        """Add a block to the block buffer.
+
+        Args:
+            block: Block to add to the block buffer.
+        """
+        self.base_batcher.add(block)
+
+    def can_add(self, block: Block) -> bool:
+        """Whether the block can be added to the buffer."""
+        return self.base_batcher.can_add(block)
+
+    def done_adding(self) -> bool:
+        """Indicate to the batcher that no more blocks will be added to the buffer."""
+        return self.base_batcher.done_adding()
+
+    def is_done_adding(self) -> bool:
+        """Whether this batcher has received an indication that
+        it should no longer receive more blocks."""
+        return self.base_batcher.is_done_adding()
+
+    def has_batch(self) -> bool:
+        """Whether this Batcher has any full batches."""
+        return self.base_batcher.has_batch() or not self.buffer.empty()
+
+    def has_any(self) -> bool:
+        """Whether this Batcher has any data."""
+        return self.base_batcher.has_any() or not self.buffer.empty()
+
+    def _prefetch_next_batch(self) -> None:
+        while self.base_batcher.has_batch() and not self.buffer.full():
+            next_batch = self.base_batcher.next_batch()
+            self.buffer.put(next_batch)
+
+        # Fetch any leftover data.
+        if (
+            not self.buffer.full()
+            and self.base_batcher.has_any()
+            and self.base_batcher.is_done_adding()
+        ):
+            next_batch = self.base_batcher.next_batch()
+            self.buffer.put(next_batch)
+
+    def next_batch(self) -> BatchType:
+        """Get the next batch from the block buffer.
+
+        Returns:
+            A formatted batch of data.
+        """
+        # If assert statement passes, then there must be at least some
+        # data in the base batcher.
+        assert (
+            not self.buffer.empty()
+            or self.base_batcher.has_batch()
+            or (self.base_batcher.is_done_adding() and self.has_any())
+        )
+
+        if self.buffer.empty() and not self.is_thread_running:
+            self._prefetch_next_batch()
+            assert not self.buffer.empty()
+
+        try:
+            batch = self.buffer.get(block=True, timeout=self.fetch_timeout_s)
+        except queue.Empty:
+            raise RuntimeError(
+                "Fetching the next batch took longer than "
+                f"{self.fetch_timeout_s} seconds"
+            )
```

Comment:
Hmm I don't think we should have a timeout defined; the existing semantics should be maintained, where we wait for as long as it takes to prepare the batch. Any reason that we'd want to add a timeout that I'm not thinking of?

Reply:
This is really meant as an emergency safety hatch to avoid infinite hanging in case something goes terribly wrong. 

The default value is already very high, but I can increase it if you think that would be better.